Initializing an array of type `void` crashes the compiler

<!--
 Thanks for submitting an issue to AssemblyScript! Please take a moment to
 read the contributing guidelines linked below to get off to a good start 🙂
-->

## Code

```ts
const e: Array<void> = [];
```

[Try it in the playground](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWIKY29uc3QgZTogQXJyYXk8dm9pZD4gPSBbXTsKIyFodG1sCjx0ZXh0YXJlYSBpZD0ib3V0cHV0IiBzdHlsZT0iaGVpZ2h0OiAxMDAlOyB3aWR0aDogMTAwJSIgcmVhZG9ubHk+PC90ZXh0YXJlYT4KPHNjcmlwdD4KbG9hZGVyLmluc3RhbnRpYXRlKG1vZHVsZV93YXNtLCB7IC8qIGltcG9ydHMgKi8gfSkKICAudGhlbigoeyBleHBvcnRzIH0pID0+IHsKICAgIGNvbnN0IG91dHB1dCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdvdXRwdXQnKQogICAgZm9yIChsZXQgaSA9IDA7IGkgPD0gMTA7ICsraSkgewogICAgICBvdXRwdXQudmFsdWUgKz0gYGZpYigke2l9KSA9ICR7ZXhwb3J0cy5maWIoaSl9XG5gCiAgICB9CiAgfSkKPC9zY3JpcHQ+)

## Stack trace

```
▌ Whoops, the AssemblyScript compiler has crashed during compile :-(
▌ 
▌ Here is the stack trace hinting at the problem, perhaps it's useful?
▌ 
▌ AssertionError: assertion failed
▌     at assert (D:\assemblyscript\std\portable\index.js:200:11)
▌     at Compiler.writeStaticBuffer (D:\assemblyscript\src\compiler.ts:2033:16)
▌     at Compiler.addStaticBuffer (D:\assemblyscript\src\compiler.ts:2044:10)
▌     at Compiler.compileArrayLiteral (D:\assemblyscript\src\compiler.ts:8393:32)
▌     at Compiler.compileLiteralExpression (D:\assemblyscript\src\compiler.ts:8066:21)
▌     at Compiler.compileExpression (D:\assemblyscript\src\compiler.ts:3519:21)
▌     at Compiler.compileGlobal (D:\assemblyscript\src\compiler.ts:1262:25)
▌     at Compiler.compileTopLevelStatement (D:\assemblyscript\src\compiler.ts:2140:20)
▌     at Compiler.compileFile (D:\assemblyscript\src\compiler.ts:1111:12)
▌     at Compiler.compile (D:\assemblyscript\src\compiler.ts:486:14)
▌     at Object.compile (D:\assemblyscript\src\index.ts:280:32)
▌     at D:\assemblyscript\cli\asc.js:902:31
▌     at measure (D:\assemblyscript\cli\asc.js:1409:3)
▌     at Object.main (D:\assemblyscript\cli\asc.js:900:24)
▌     at runTest (D:\assemblyscript\tests\compiler.js:178:7)
▌ 
▌ If it refers to the dist files, try to 'npm install source-map-support' and
▌ run again, which should then show the actual code location in the sources.
▌ 
▌ If you see where the error is, feel free to send us a pull request. If not,
▌ please let us know: https://github.com/AssemblyScript/assemblyscript/issues
▌ 
▌ Thank you!
```

--- a/src/compiler.ts
+++ b/src/compiler.ts
@@ -39,6 +39,7 @@ import {
   getConstValueI64High,
   getConstValueF32,
   getConstValueF64,
+  getConstValueV128,
   getBlockChildCount,
   getBlockChildAt,
   getBlockName,
@@ -193,6 +194,7 @@ import {
   writeI64,
   writeF32,
   writeF64,
+  writeV128,
   uniqueMap,
   isPowerOf2,
   v128_zero,
@@ -1927,6 +1929,20 @@ export class Compiler extends DiagnosticEmitter {
           writeF64(getConstValueF64(value), buf, pos);
           pos += 8;
         }
+        break;
+      }
+      case <u32>TypeRef.V128: {
+        for (let i = 0; i < length; ++i) {
+          let value = values[i];
+          assert(getExpressionType(value) == elementTypeRef);
+          assert(getExpressionId(value) == ExpressionId.Const);
+          writeV128(getConstValueV128(value), buf, pos);
+          pos += 16;
+        }
+        break;
+      }
+      case <u32>TypeRef.None: {
+        // nothing to write
         break;
       }
       default: assert(false);
@@ -8269,11 +8285,15 @@ export class Compiler extends DiagnosticEmitter {
       let elementExpression = expressions[i];
       if (elementExpression.kind != NodeKind.OMITTED) {
         let expr = this.compileExpression(<Expression>elementExpression, elementType, Constraints.CONV_IMPLICIT);
-        let precomp = module.runExpression(expr, ExpressionRunnerFlags.PreserveSideeffects);
-        if (precomp) {
-          expr = precomp;
-        } else {
+        if (getExpressionType(expr) != elementType.toRef()) {
           isStatic = false;
+        } else {
+          let precomp = module.runExpression(expr, ExpressionRunnerFlags.PreserveSideeffects);
+          if (precomp) {
+            expr = precomp;
+          } else {
+            isStatic = false;
+          }
         }
         values[i] = expr;
       } else {

--- a/src/module.ts
+++ b/src/module.ts
@@ -2665,6 +2665,17 @@ export function getConstValueF64(expr: ExpressionRef): f64 {
   return binaryen._BinaryenConstGetValueF64(expr);
 }
 
+export function getConstValueV128(expr: ExpressionRef): Uint8Array {
+  let cArr = binaryen._malloc(16);
+  binaryen._BinaryenConstGetValueV128(expr, cArr);
+  let out = new Uint8Array(16);
+  for (let i = 0; i < 16; ++i) {
+    out[i] = binaryen.__i32_load8_u(cArr + i);
+  }
+  binaryen._free(cArr);
+  return out;
+}
+
 export function isConstZero(expr: ExpressionRef): bool {
   if (getExpressionId(expr) != ExpressionId.Const) return false;
   var type = getExpressionType(expr);

--- a/src/util/binary.ts
+++ b/src/util/binary.ts
@@ -87,3 +87,14 @@ export function writeF64(value: f64, buffer: Uint8Array, offset: i32): void {
   writeI32(i64_low(valueI64), buffer, offset);
   writeI32(i64_high(valueI64), buffer, offset + 4);
 }
+
+/** Reads a 128-bit vector from the specified buffer. */
+export function readV128(buffer: Uint8Array, offset: i32): Uint8Array {
+  return buffer.slice(offset, offset + 16);
+}
+
+/** Writes a 128-bit vector to the specified buffer. */
+export function writeV128(value: Uint8Array, buffer: Uint8Array, offset: i32): void {
+  assert(value.length == 16);
+  buffer.set(value, offset);
+}

